### PR TITLE
fix: Fix stepper charge hypothesis

### PR DESCRIPTION
**Pull request by andiwand**
I found at least one case where we rely on the sign of the charge hypothesis in the stepper. The charge might flip during KF but the one inside the stepper would stay the same causing inconsistencies and in my case a terrible fit.

In this PR I remove the sign of the charge hypothesis, refactored the `momentum` and `charge` accessor methods and added a `qop` accessor method that should be preferred over explicit `charge / momentum` IMO

### CODE REVIEW

File: Core/include/Acts/Propagator/EigenStepper.hpp [hunk 1]
```diff
@@ -213,17 +213,24 @@ class EigenStepper {
...
+  ///
+  /// @param state [in] The stepping state (thread-local cache)
+  double qop(const State& state) const { return state.pars[eFreeQOverP]; }
+
   /// Absolute momentum accessor
   ///
   /// @param state [in] The stepping state (thread-local cache)
   double momentum(const State& state) const {
-    return std::abs((state.q == 0. ? 1. : state.q) / state.pars[eFreeQOverP]);
+    return std::abs(charge(state) / qop(state));
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/EigenStepper.hpp:225]: What happens if `q==0`?

File: Core/include/Acts/Propagator/EigenStepper.hpp [hunk 2]
```diff
@@ -213,17 +213,24 @@ class EigenStepper {
...
   }
 
   /// Charge access
   ///
   /// @param state [in] The stepping state (thread-local cache)
-  double charge(const State& state) const { return state.q; }
+  double charge(const State& state) const {
+    return std::copysign(state.absCharge, qop(state));
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/EigenStepper.hpp:232]: Charge can't be zero this way can it?

Comment (paulgessinger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:232]: Ah, theres `+0` and `-0`, right.

File: Core/include/Acts/Propagator/EigenStepper.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/Propagator/EigenStepper.hpp:225]: Charge will be `+-0`, but you take `abs`, so it's `|p|`, which is all good.

Comment (andiwand) [reply Core/include/Acts/Propagator/EigenStepper.hpp:225]: oh yeah I have to catch the q=0 case

File: Core/include/Acts/Propagator/EigenStepper.ipp [hunk 3]
```diff
@@ -84,12 +84,12 @@ void Acts::EigenStepper<E, A>::update(State& state,
 
 template <typename E, typename A>
 void Acts::EigenStepper<E, A>::update(State& state, const Vector3& uposition,
-                                      const Vector3& udirection, double up,
+                                      const Vector3& udirection, double qop,
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/EigenStepper.ipp:87]: I'm guessing you checked all uses of this function to make sure we're passing in the right value?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp [hunk 4]
```diff
@@ -63,23 +63,33 @@ struct WeightedComponentReducerLoop {
...
   }
 
+  template <typename stepper_state_t>
+  static ActsScalar qop(const stepper_state_t& s) {
+    return std::accumulate(
+        s.components.begin(), s.components.end(), ActsScalar{0.},
+        [](const auto& sum, const auto& cmp) -> ActsScalar {
+          return sum + cmp.weight * cmp.state.pars[eFreeQOverP];
+        });
+  }
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:73]: Should this be cached and updated when changes happen? (/cc @benjaminhuth)

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp [hunk 5]
```diff
@@ -63,23 +63,33 @@ struct WeightedComponentReducerLoop {
     return toVector3(s.components, eFreeDir0).normalized();
   }
 
+  template <typename stepper_state_t>
```
Comment (benjaminhuth) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:66]:
```suggestion
  // TODO: Maybe we can cache this value and only update it when the parameters change
  template <typename stepper_state_t>
```

File: Core/include/Acts/Propagator/EigenStepper.ipp (hunk 3 above)
Comment (andiwand) [reply Core/include/Acts/Propagator/EigenStepper.ipp:87]: I checked again and it seems to be only used in `PointwiseMaterialInteraction` and `SimulationActor`

Comment (paulgessinger) [reply Core/include/Acts/Propagator/EigenStepper.ipp:87]: Don't we use it in the CKF too?

Comment (andiwand) [reply Core/include/Acts/Propagator/EigenStepper.ipp:87]: only through `PointwiseMaterialInteraction`. the `CKF` is using the other overloads

Comment (paulgessinger) [reply Core/include/Acts/Propagator/EigenStepper.ipp:87]: ok

**Review by asalzburger — APPROVED** [review]
That looks all consistent, good to go.

**Review by asalzburger — APPROVED** [review]
Re-approving

**Comment by paulgessinger** [comment]
Just for reference, there are some discussions that this might have improved measurably the ACTS CKF performance as run in Athena (https://its.cern.ch/jira/browse/ATLIDTRKCP-472)

**Comment by andiwand** [comment]
> Just for reference, there are some discussions that this might have improved measurably the ACTS CKF performance as run in Athena (https://its.cern.ch/jira/browse/ATLIDTRKCP-472)

Noice! I also saw improvements in our ODD full chain but it was mostly outliers

**Pull request merged**